src/radix_connect/mobile/client.rs

```diff
@@ -1,65 +1,680 @@
+        Ok(RadixConnectMobileSessionRequest::new(
+            request.session_id,
+            request.interaction,
+            request.origin,
+            origin_requires_validation,
+        ))
+    }
+
+    /// Send the Wallet's response to the dApp.
+    /// This is a stateful operation as it will save the session in the secure storage if the user validated the session.
+    #[uniffi::method]
+    pub async fn send_dapp_interaction_response(
+        &self,
+        wallet_response: RadixConnectMobileWalletResponse,
+    ) -> Result<()> {
+        let session_id = wallet_response.session_id;
+
+        // Get the in flight session, if any, that required validation from the user.
+        // Not remove at this stage, as the user might retry the interaction.
+        let in_flight_session = self
+            .new_sessions
+            .read()
+            .ok()
+            .and_then(|sessions| sessions.get(&session_id).cloned());
+
+        // Get the existing session, if any.
+        let existing_session =
+            self.load_session(wallet_response.session_id).await.ok();
+
+        let is_in_flight_session = in_flight_session.is_some();
+        let session = in_flight_session.or(existing_session).ok_or(
+            CommonError::RadixConnectMobileSessionNotFound { session_id },
+        )?;
+
+        let is_success_response = wallet_response.response.is_success();
+
+        // Send the wallet interaction response to the dApp through the transport.
+        self.wallet_interactions_transport
+            .send_wallet_interaction_response(
+                session.clone(),
+                wallet_response.response,
+            )
+            .await?;
+
+        _ = self.new_sessions.try_write().map(|mut new_sessions| {
+            new_sessions.remove(&session_id);
+        });
+
+        if is_in_flight_session && is_success_response {
+            // We do consider a session to be validated once user did send a successful interaction back.
+            self.save_session(session).await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl RadixConnectMobile {
+    const HKDF_KEY_DERIVATION_INFO: &'static str = "RCfM";
+
+    fn create_in_flight_session(
+        &self,
+        request: &RadixConnectMobileRequest,
+    ) -> Result<()> {
+        // 1. Generate the Wallet's private/public key pair
+        let wallet_private_key = KeyAgreementPrivateKey::generate()?;
+        let wallet_public_key = wallet_private_key.public_key();
+
+        // 2. Generate the secret that is shared by the Wallet and the dApp
+        let shared_secret = wallet_private_key
+            .shared_secret_from_key_agreement(&request.public_key);
+
+        // 3. Derive the encryption key from the shared secret
+        let dapp_definition_address = request.dapp_definition_address.address();
+        let encryption_key = PbHkdfSha256::hkdf_key_agreement(
+            shared_secret.as_bytes(),
+            Some(dapp_definition_address.as_bytes()),
```

Comment:
can drop `Some(` if you do the suggested `impl Into<Option<T>>` I suggested above for `PbHkdfSha256` function..